File: src/main/java/com/sidgate/poker/Straight.java
package com.sidgate.poker;

import com.sidgate.deck.Card;
import com.sidgate.deck.Hand;

import java.util.LinkedList;

public class Straight implements HandType {
	@Override
	public boolean validate(Hand hand) {
		LinkedList<Card> sorted = hand.sortedWithAceAdjusted();
		Card previous = null;
		for (Card card : sorted) {
			if (previous != null) {
				int previousRank = previous.isAce() ? -1 : previous.rank().ordinal();
				if (card.rank().ordinal() - previousRank != 1) {
					return false;
				}
			}
			previous = card;
		}
		return true;
	}

	@Override
	public String getName() {
		return "straight";
	}
}

File: src/main/kotlin/sidgate/poker/Straight.kt
package sidgate.poker

import sidgate.deck.Card
import sidgate.deck.Hand

class Straight(next: HandType?) : HandType(next) {
    override fun process(hand: Hand): Boolean {
        val sorted = hand.sortedWithAceAdjusted()
        var previous: Card? = null
        for (card in sorted) {
            if (previous != null) {
                val previousRank = if (previous.isAce) -1 else previous.rank.ordinal
                if (card.rank.ordinal - previousRank != 1) {
                    return false
                }
            }
            previous = card
        }
        return true
    }

    override val name: String
        get() = "straight"
}